### DISCUSSION TITLE
过期：HDR 转 SDR 的优化方案

### DISCUSSION THREAD

**Discussion by natural-harmonia-gropius** [hooke007/mpv_PlayKit#25] [回收]
**本文以 `vo=gpu-next` 为必要条件**

目前的默认参数我认为没什么可调整的，也可以参考下面的讨论
如果觉得亮度总是在跳动那就在 mpv.conf 中加一行 `hdr-compute-peak=no`

## Shader

但是我个人并不满意 mpv 的映射结果，所以自己做了转换着色器
https://github.com/natural-harmonia-gropius/hdr-toys

缺点是字幕会从 srgb 转成 bt.2100-pq 非常暗淡

## LUT

**目前 lut 不能在运行时改变**

从达芬奇导出的 lut，和一个 [Toru Yoshihara](https://github.com/toru-ver4) 从 youtube 反推的 lut，可作为输出参考
https://github.com/natural-harmonia-gropius/mpv_config/tree/master/portable_config/luts

```ini
target-trc=pq
target-prim=bt.2020
target-lut=~~/luts/DaVinci_Rec.2020_ST2084_to_Rec.709_Gamma2.4.cube
```

**Comment by hooke007** [comment]
`icc-profile-auto`

引入的变量太大了，而且作为个人校色的结果基本都是错的（这一点在现有的流行科普里并没有提）。
折腾一圈最后放弃了icc这条扭曲的道路
我现在的做法是放弃校色使用经过专业设备和人员修正过的srgb色域显示器

**Reply by hooke007** [reply]
涉及到icc和hdr我还是先搁置一下吧，水太深而且好多都没定论（比如mpv hdr2sdr后的gamma是否 [过曝](https://github.com/mpv-player/mpv/issues/8009) // 一些厂商的标准色温正在由d65向d68过渡（这在不少测评里被习惯性的认为色温做错了。。。） // 以及 [这个问题](https://github.com/mpv-player/mpv/issues/9173) 甚至让我对当前普遍的sdr源的处理是不是正确都怀疑），已手动置顶该讨论，如果他人看见说不定可以尝试一下，有用。

**Reply by hooke007** [reply]
话说hdr转sdr后再套你新传的ACES不会产生二次映射吗，我一直没找到禁止mpv内部映射的方法

**Reply by natural-harmonia-gropius** [reply]
肯定会二次映射，所以说不正确。第一次套完着色器也是不正确的，套完就应该视作gamma曲线来处理。
目前无法禁止mpv内部映射，所以无解。
就当套了个去雾滤镜吧 😅

**Reply by hooke007** [reply]
。。。那vo_gpu依然是无解了

**Comment by natural-harmonia-gropius** [comment]


**Reply by dyphire** [reply]
我也发现uwp的电影电视hdr映射表现不错，不过我这里测试效果并没有madvr表现好。~~顺便--target-peak=100虽然对提升暗部细节有帮助，但高光极易过曝，我最终还是选择使用默认203。~~ 发现--target-peak参数的高光表现是受--hdr-compute-peak和--tone-mapping-desaturate-exponent参数的影响，--no-hdr-compute-peak或者tone-mapping-desaturate-exponent=0都可以尽可能地避免高光过曝。mpv的hdr映射确实比madvr要差一截

**Reply by natural-harmonia-gropius** [reply]
我更在意的一个问题是亮度改变时偏色，亮度越高cyan、yellow、magenta会在hue中占据越大的turn（之前在mpv的issue下抱怨过的黄色变橙色，还有紫罗兰那个地面发绿，我觉得正确颜色应该是cyan）

**Reply by dyphire** [reply]
偏色这问题我也发现了，明暗对比度的调整也的确难以找到满意的方案。期待新参数等着抄作业（

**Comment by zhongfly** [comment]
我问个略微离题的问题:在普通显示器上(不支持HDR)，播放HDR视频是否有意义？或者说观感与非hdr相比有提升吗？

**Reply by hooke007** [reply]
有些电影原盘4k就只有hdr，就想看4k高分的，就只有tonemapping可选了

**Reply by natural-harmonia-gropius** [reply]
摄像机一类的采集到的log或者raw都可以视作hdr，那些所谓的sdr其实也是tone-mapping下来的产物

**Reply by zhongfly** [reply]
> 摄像机一类的采集到的log或者raw都可以视作hdr，那些所谓的sdr其实也是tone-mapping下来的产物

这么说来假如有2个视频只有HDR和SDR的区别，更建议在普通设备上观看SDR版本的？

**Reply by hooke007** [reply]
选择 厂商预设调好的tonemapping方案 还是 我自己调校的tonemapping方案。我倾向前者。
更像是一个主观偏好问题而不是客观正确性的问题。

_上纲上线的话也可以说官方sdr是导演意图，我们hdr2sdr就是瞎折腾（_

**Reply by natural-harmonia-gropius** [reply]
电视厂赞助的片和胶卷翻录除外，这两类选HDR

**Comment by natural-harmonia-gropius** [comment]
动态曲线初步可用，有兴趣的可以试试
https://github.com/Natural-Harmonia-Gropius/mpv_config/pull/10

**Reply by hooke007** [reply]
我没仔细看怎么实现的，大致浏览了一下，不清楚next对前后帧的数据是怎么获取的。通过运行时不断传递新的glsl-shader-opt，最大的问题是怎么在下一帧呈现前完成着色器的重新编译和渲染，这在时间轴上是否能匹配的上。另外再考虑到不同机器的性能问题。。。

**Reply by natural-harmonia-gropius** [reply]
亮度是从当前帧实时获取的，参数也是在同一帧计算的
hdr-toys-helper.lua 传的 L_hdr 一个文件只会有一次，这里主要为了 chroma_correction_hsv 高光降饱和跟随静态元信息走，piecewise 没有用到这个参数

性能问题单纯是我太菜，用上 compute shader 再优化一下管线应该能提升好几倍

前后帧 nlmeans 用的 //!STORAGE 这个我还没了解

**Reply by natural-harmonia-gropius** [reply]
[动态曲线](https://github.com/Natural-Harmonia-Gropius/mpv_config/pull/10)的性能问题已经解决，闪烁也极大改善了。

libplacebo 的 [st2094](https://code.videolan.org/videolan/libplacebo/-/merge_requests/360) 动态曲线也快上线了

**Reply by dyphire** [reply]
我也注意到libplacebo那个pr了

**Reply by dyphire** [reply]
shader 动态映射方案的画面表现相当不错了，但是字幕颜色的扭曲效果仍然无法接受
| gpu-next bt.2446a                                            | hable2 dynamic hybrid                                        |
| ------------------------------------------------------------ | ------------------------------------------------------------ |
| ![](https://cdn.jsdelivr.net/gh/dyphire/PicGo/img/2023/02/08/20230208203107.png) | ![](https://cdn.jsdelivr.net/gh/dyphire/PicGo/img/2023/02/08/20230208203150.png) |

shader 的字幕色彩已经面目全非了，还是等libplacebo的st2094动态映射曲线的原生实现

**Reply by natural-harmonia-gropius** [reply]
字幕也从srgb转成2020 pq了，需要mpv为字幕添加设置项，就是不知道他们接受不接受这种奇怪的请求了

**Reply by dyphire** [reply]
shader 动态映射的色彩准确度有很大问题啊

| gpu-next bt.2446a                                            | hable2 dynamic hybrid                                        |
| ------------------------------------------------------------ | ------------------------------------------------------------ |
| ![](https://cdn.jsdelivr.net/gh/dyphire/PicGo/img/2023/02/09/20230209121625.png) | ![](https://cdn.jsdelivr.net/gh/dyphire/PicGo/img/2023/02/09/20230209120918.png) |

**Reply by natural-harmonia-gropius** [reply]
这个是对的
jch中h直接使用原画面计算，jc混合了max的j和rgb的c所以高亮度的色彩会衰减，其中c的衰减应该和ictcp降低ctcp得到的效果一致
暗部因为最亮处为10000nit，相比起203nit提升了5.6档，默认1000：1的对比度，结果就是低于5.6*203/1000nit的地方都是曲线的toe，相较暗部是线性的曲线会更暗
亮部还有一部分原因是chroma_correction，默认的peak是1000，而这个视频的元信息写的10000，你需要 [hdr-toys-helper.lua](https://github.com/Natural-Harmonia-Gropius/mpv_config/blob/main/portable_config/scripts/hdr-toys-helper.lua) 辅助自动完成这一步

**Comment by yeezylife** [comment]
![Terminator 2 Judgment Day 1991 1080p UHD BluRay DD+5 1 HDR x265-CtrlHD- 01_20_24 278](https://user-images.githubusercontent.com/30175425/215075833-c91cd178-a1d8-4a42-88c3-f0d9e35a9b78.jpg)
用了你的配置文件 和 shaders，感觉画面变得有点暗（只是个人感觉）。 
更重要的是有些画面上会有迷之黑点。。

**Reply by natural-harmonia-gropius** [reply]
片源能提供一下吗

**Reply by natural-harmonia-gropius** [reply]
我忽然发现这里的链接是老的，你现在用的是这个吗
https://github.com/Natural-Harmonia-Gropius/mpv_config/tree/main/portable_config/shaders/hdr-toys

**Reply by yeezylife** [reply]
我是在你的github里下的最新的。

我需要指出的是我只测试过这一个HDR片。。是我自己下错版本了，于是才开始尝试HDR转SDR

至于出现频率，这个片我现在看了一个半小时出现过四五次这种黑点了。

**Reply by natural-harmonia-gropius** [reply]
公网没找到，用 ffmpeg 截一段给我看看……
`ffmpeg -i input.mp4 -ss 00:00 -t 10 output.mp4`

还有 `shift+i` `2` 的截图是啥样的
![image](https://user-images.githubusercontent.com/50797982/215079969-40b03169-adb2-4827-83c6-e00508b22eed.png)

**Comment by yeezylife** [comment]
![Terminator 2 Judgment Day 1991 1080p UHD BluRay DD+5 1 HDR x265-CtrlHD- 01_20_22 609](https://user-images.githubusercontent.com/30175425/215080270-d5145ac3-ea4e-4565-9145-5fb1d84b5f08.jpg)
我先上shift+i 2 吧 看着是生效了啊 有你那些shaders

截段等等 对我来说这是技术活 我得研究

**Comment by yeezylife** [comment]
![Terminator 2 Judgment Day 1991 1080p UHD BluRay DD+5 1 HDR x265-CtrlHD- 00_00_02 294](https://user-images.githubusercontent.com/30175425/215081125-f3c79936-e229-4ff6-9ff8-0c10c72bcfcc.jpg)

我删去自己所有的设置 只留下 vo=gpu-next 变成了黑色块

出现黑点/色块的地方 可以反复复现！

**Reply by natural-harmonia-gropius** [reply]
我这无法复现，完全使用我的设置会有色块吗

![image](https://user-images.githubusercontent.com/50797982/215081887-0eee439e-eb22-4e5a-9fa0-a0c89d5ebf42.png)

**Comment by yeezylife** [comment]
```#Decode
vo=gpu-next
gpu-api=d3d11
# target-colorspace-hint

[bt.2020]
profile-cond=get("video-params/primaries") == "bt.2020" and get("video-params/sig-peak") == 1
profile-restore=copy
target-prim=bt.2020
glsl-shader=~~/shaders/hdr-toys/transfer-function/bt1886_to_linear.glsl
glsl-shader=~~/shaders/hdr-toys/gamut-mapping/compress.glsl
glsl-shader=~~/shaders/hdr-toys/transfer-function/linear_to_bt1886.glsl

[bt.2100]
profile-cond=get("video-params/sig-peak") > 1
profile-restore=copy
target-trc=pq
target-prim=bt.2020
glsl-shader=~~/shaders/hdr-toys/utils/clip_both.glsl
glsl-shader=~~/shaders/hdr-toys/transfer-function/pq_to_l.glsl
glsl-shader=~~/shaders/hdr-toys/transfer-function/l_to_linear.glsl
glsl-shader=~~/shaders/hdr-toys/utils/crosstalk.glsl
glsl-shader=~~/shaders/hdr-toys/utils/chroma_correction.glsl
glsl-shader=~~/shaders/hdr-toys/tone-mapping/bt2446c.glsl
glsl-shader=~~/shaders/hdr-toys/utils/crosstalk_inverse.glsl
glsl-shader=~~/shaders/hdr-toys/gamut-mapping/compress.glsl
glsl-shader=~~/shaders/hdr-toys/transfer-function/linear_to_bt1886.glsl


这是我全部的配置文件了 一样有黑色块。。。奇了怪了

**Reply by natural-harmonia-gropius** [reply]
可以复现了，用 d3d11 会有这个问题，vulkan 没问题

**Reply by yeezylife** [reply]
vulkan 在win上N卡驱动有问题 还剩下最后半小时我凑合看完算了。。

**Reply by natural-harmonia-gropius** [reply]
亮度的话你试试 `set glsl-shader-opts L_sdr=100`

**Reply by natural-harmonia-gropius** [reply]
chroma_correction 那一行去掉就没有黑点了

**Reply by yeezylife** [reply]
确实去掉chroma_correction 那一行 就没了。。。这能随便去的么 不会影响效果么
你发现的好快呀

**Reply by natural-harmonia-gropius** [reply]
对这个片子，我觉得是不影响效果的……

**Reply by natural-harmonia-gropius** [reply]
修好了，我这边测试没问题了

**Comment by dyphire** [comment]
HDR 映射上我个人目前还是选择 mpv 的内部映射算法，主要是上游近期已修复映射过饱和问题。shader 着色器映射方案虽然可玩性不小，不过 OSD 映射上的缺失还是痛点（mpv 本身的问题）

而且目前的 shader 组合方案测试下来也还不是很完善，mpv 的内部映射算法个人测试下效果最好的还是 --vo=gpu-next 下的 bt.2446a 映射曲线。~~不过 bt.2446a 默认参数映射画面较暗，只能搭配 target 色彩管理方案使用（icc 下目前没有好的处理方式）~~

Edit: 发现 hdr-compute-peak 对 bt.2446a 的对比度有明显改善，恢复默认值。dovi 的 p5 无相关亮度峰值信息，可根据喜好自行为 p5 切换 hable 色调映射算法

个人的 HDR to SDR 映射方案的相关参数参考
```ini
vo=gpu-next
gpu-api=d3d11
target-prim=auto
target-trc=gamma2.2
target-peak=auto

tone-mapping=bt.2446a
tone-mapping-crosstalk=0.04
tone-mapping-mode=auto
gamut-mapping-mode=clip
hdr-compute-peak=auto

[HDR]
profile-desc=HDR配置参数
profile-cond=p["video-params/sig-peak"] > 1
profile-restore=copy
#target-peak=100
tone-mapping-mode=rgb
blend-subtitles=no
deband=no
vf=""

[hdr-spline]
profile-cond=p.tone_mapping == "spline" and p["video-params/sig-peak"] > 1000 / 203
profile-restore=copy
tone-mapping-param=0.15

[hdr-2390]
profile-cond=p.tone_mapping == "bt.2390" and p.current_vo == "gpu-next"
profile-restore=copy
tone-mapping-param=2.0

[hdr-peak]
profile-cond=p.tone_mapping == "reinhard" or p.tone_mapping == "bt.2390"
profile-restore=copy
hdr-compute-peak=no
```

**Reply by natural-harmonia-gropius** [reply]
next 的红色的高光和阴影依然还是有问题。
| gpu-next   | shader  |
| ---------- | ----------- |
|![image](https://user-images.githubusercontent.com/50797982/215322338-5793ee00-a7eb-4388-9404-f707a314af21.png)|![image](https://user-images.githubusercontent.com/50797982/215322335-6f9a19e9-3145-4ea8-82c7-c282b08f6ff5.png)|

另外 next 的 1886 的黑位始终是 1/对比度，我这边除了少数几个地方都为 0，这也是 bt2020 转 bt709 next 的 clip 比我这边 matrix 更淡一些的原因

**Reply by dyphire** [reply]
> next 的红色的高光和阴影依然还是有问题

可能跟 --tone-mapping-crosstalk 参数有关，我测试中它会影响高光下的色准

顺便我又测试了下你最新的 shader 着色器映射方案，表现确实很不错了，之前的旧印象已经过时了。不过它除了 OSD 外还会影响到字幕渲染就让我很难接受了
| gpu-next bt.2446a                                            | shader                                       |
| ------------------------------------------------------------ | ------------------------------------------------------------ |
| ![](https://cdn.jsdelivr.net/gh/dyphire/PicGo/img/2023/01/29/20230129202206.png) | ![](https://cdn.jsdelivr.net/gh/dyphire/PicGo/img/2023/01/29/20230129201947.png) |

单就画面效果感觉 shader 映射方案表现要更好，但字幕渲染就不太行了

**Reply by yeezylife** [reply]
你的这个方案比较方便，不用下着色器之类的，但是修改的设置有点多啊，尤其是这几条
```
tone-mapping=bt.2446a
tone-mapping-crosstalk=0.04
gamut-mapping-mode=clip


播放普通的SDR文件是不是不能用这些设置啊？他那个shader方案好在不用改其他设置，碰到HDR自动启用

**Reply by dyphire** [reply]
tone-mapping*和gamut-mapping-mode等参数只在hdr和色域超出bt.709(target-prim默认值)时生效，播放常规色域的SDR不会触发。建议多研究下mpv相关参数
注：在stats统计信息的第二页可以看到有没有触发tone-mapping和gamut-mapping

顺便之所以使用了条件配置切换tone-mapping-mode的参数，是因为tone-mapping-mode在非rgb模式下会过饱和hdr画面，但在dovi p5下rgb模式又会扭曲高光颜色的映射。所以通过条件配置处理

**Comment by dyphire** [comment]
刚才测试[avs3 10bit补丁](https://github.com/dyphire/mpv-winbuild-cmake/commit/04b76e135d90c19a8bf9c63b8945004c705f68fe)的[编译版](https://github.com/dyphire/mpv-winbuild/releases/tag/mpv_own-2023-01-31)时，发现 shader 映射方案在处理具有极高峰值的画面时会过曝，画面表现很像mpv内部的mobius映射曲线。

| shader                                                       | bt.2446a                                                     | bt.2390                                                      |
| ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
| ![](https://cdn.jsdelivr.net/gh/dyphire/PicGo/img/2023/01/31/20230131150324.png) | ![](https://cdn.jsdelivr.net/gh/dyphire/PicGo/img/2023/01/31/20230131155110.png) | ![](https://cdn.jsdelivr.net/gh/dyphire/PicGo/img/2023/01/31/20230131150521.png) |


测试视频：
[HDR-JNDNet_Test_qp37_avs3_HDR.zip](https://github.com/hooke007/MPV_lazy/files/10543821/HDR-JNDNet_Test_qp37_avs3_HDR.zip)

**Reply by dyphire** [reply]
可以看到 bt.2446a 高光细节保留的真不错，但是对比度表现就不如bt.2390了

**Reply by natural-harmonia-gropius** [reply]
2446c是这样的，对于这种的建议换hable2

**Reply by natural-harmonia-gropius** [reply]
忽然想到对这个视频还需要这个 https://github.com/Natural-Harmonia-Gropius/mpv_config/blob/main/portable_config/scripts/hdr-toys-helper.lua
对2446c没效果，对其他的reinhard、hable2之类的结果会有点区别

**Reply by natural-harmonia-gropius** [reply]
现在应该没这个问题了，云的细节还得等动态曲线
![image](https://user-images.githubusercontent.com/50797982/216826582-2b5fa8a1-83c6-4ec1-9152-d578b26fdd74.png)

**Reply by dyphire** [reply]
还是有点过曝，高光细节都丢了。云和工具包比较明显，雪山也是。但观感好了不少

**Comment by natural-harmonia-gropius** [comment]
@hooke007 能帮我清屏吗？我这里没有权限。除了[这一楼](https://github.com/hooke007/MPV_lazy/discussions/25#discussioncomment-4802790)其他的都隐藏掉或删掉，麻烦了。
这帖子从发出后改变太多了

**Reply by hooke007** [reply]
桌面端的渲染太蠢了，移动端折叠后子楼层也会自动折叠。我急着出门，对话太多可能需要很长时间藏干净，或者你可以考虑开新帖

**Reply by natural-harmonia-gropius** [reply]
我开那新的去了
